signer/xpi/xpi.go

```diff
@@ -143,54 +148,116 @@ func (s *PKCS7Signer) Config() signer.Configuration {
 	)
-	manifest, sigfile, err := makeJARManifests(input)
+
+	opt, err = GetOptions(options)
+	if err != nil {
+		return nil, errors.Wrap(err, "xpi: cannot get options")
+	}
+	cn, err := opt.CN(s)
+	if err != nil {
+		return nil, err
+	}
+	coseSigAlgs, err = opt.Algorithms()
+	if err != nil {
+		return nil, errors.Wrap(err, "xpi: error parsing COSE Algorithms from options")
+	}
+
+	manifest, err = makeJARManifest(input)
+	if err != nil {
+		return nil, errors.Wrap(err, "xpi: cannot make JAR manifest from XPI")
+	}
+
+	// don't issue COSE signatures if we cannot parse valid and
+	// supported COSE Algorithms from signing request options
+	if len(coseSigAlgs) < 1 {
+		pkcs7Manifest = manifest
+	} else {
+		coseSig, err = s.issueCOSESignature(cn, manifest, coseSigAlgs)
+		if err != nil {
+			return nil, errors.Wrap(err, "xpi: error signing cose message")
+		}
+
+		// add the cose files to the metafiles we'll add to the XPI
+		coseMetaFiles := []Metafile{
+			{"META-INF/cose.manifest", manifest},
+			{"META-INF/cose.sig", coseSig},
+		}
+		metas = append(metas, coseMetaFiles...)
+
+		// add entries for the cose files to the manifest as cose.manifest and cose.sig
+		pkcs7Manifest, err = makePKCS7Manifest(input, metas)
+		if err != nil {
+			return nil, errors.Wrap(err, "xpi: error making PKCS7 manifest")
+		}
+	}
+
+	sigfile, err := makeJARSignature(pkcs7Manifest)
 	if err != nil {
-		return nil, errors.Wrap(err, "xpi: cannot make JAR manifests from XPI")
+		return nil, errors.Wrap(err, "xpi: cannot make JAR manifest signature from XPI")
 	}
-	p7sig, err := s.signData(sigfile, options)
+
+	p7sig, err := s.signDataWithPKCS7(sigfile, cn)
 	if err != nil {
 		return nil, errors.Wrap(err, "xpi: failed to sign XPI")
 	}
-	signedFile, err = repackJAR(input, manifest, sigfile, p7sig)
+
+	metas = append(metas, []Metafile{
+		{"META-INF/manifest.mf", pkcs7Manifest},
+		{"META-INF/mozilla.sf", sigfile},
+		{"META-INF/mozilla.rsa", p7sig},
+	}...)
+
+	signedFile, err = repackJARWithMetafiles(input, metas)
 	if err != nil {
 		return nil, errors.Wrap(err, "xpi: failed to repack XPI")
 	}
 	return signedFile, nil
 }
 
-// SignData takes an input signature file and returns a PKCS7 detached signature
-func (s *PKCS7Signer) SignData(sigfile []byte, options interface{}) (signer.Signature, error) {
-	p7sig, err := s.signData(sigfile, options)
+// SignData takes an input signature file and returns a PKCS7 or COSE detached signature
+func (s *XPISigner) SignData(sigfile []byte, options interface{}) (signer.Signature, error) {
+	opt, err := GetOptions(options)
+	if err != nil {
+		return nil, errors.Wrap(err, "xpi: cannot get options")
+	}
+	cn, err := opt.CN(s)
+	if err != nil {
+		return nil, err
+	}
+	coseSigAlgs, err := opt.Algorithms()
+	if err != nil {
+		return nil, errors.Wrap(err, "xpi: cannot make JAR manifest from XPI")
+	}
+	if len(coseSigAlgs) > 0 {
+		return nil, fmt.Errorf("xpi: cannot use /sign/data for COSE signatures. Use /sign/file instead")
```

Comment:
The error msg above is wrong. Julien wants to only issue COSE signatures from `/sign/file` to avoid breaking `/sign/data` which is the only endpoint in use. Updated to skip parsing and validating the `cose_algorithms` from the JSON entirely and just check for a non-empty array.